core/dataset_next.h

```diff
@@ -0,0 +1,586 @@
+  auto begin() & noexcept {
+    return boost::make_transform_iterator(m_data.begin(),
+                                          detail::make_item{this});
+  }
+  auto end() const && = delete;
+  auto end() && = delete;
+  /// Return const iterator to the end of all data items.
+  auto end() const &noexcept {
+    return boost::make_transform_iterator(m_data.end(),
+                                          detail::make_item{this});
+  }
+  /// Return iterator to the end of all data items.
+  auto end() & noexcept {
+    return boost::make_transform_iterator(m_data.end(),
+                                          detail::make_item{this});
+  }
+
+  void setCoord(const Dim dim, Variable coord);
+  void setLabels(const std::string &labelName, Variable labels);
+  void setAttr(const std::string &attrName, Variable attr);
+  void setValues(const std::string &name, Variable values);
+  void setVariances(const std::string &name, Variable variances);
+  void setSparseCoord(const std::string &name, Variable coord);
+  void setSparseLabels(const std::string &name, const std::string &labelName,
+                       Variable labels);
+
+  DatasetConstProxy slice(const Slice slice1) const;
+  DatasetConstProxy slice(const Slice slice1, const Slice slice2) const;
+  DatasetConstProxy slice(const Slice slice1, const Slice slice2,
+                          const Slice slice3) const;
+  DatasetProxy slice(const Slice slice1);
+  DatasetProxy slice(const Slice slice1, const Slice slice2);
+  DatasetProxy slice(const Slice slice1, const Slice slice2,
+                     const Slice slice3);
+
+  bool operator==(const Dataset &other) const;
+  bool operator==(const DatasetConstProxy &other) const;
+  bool operator!=(const Dataset &other) const;
+  bool operator!=(const DatasetConstProxy &other) const;
+
+private:
+  friend class DatasetConstProxy;
+  friend class DatasetProxy;
+  friend class DataConstProxy;
+  friend class DataProxy;
+
+  void setExtent(const Dim dim, const scipp::index extent, const bool isCoord);
+  void setDims(const Dimensions &dims, const Dim coordDim = Dim::Invalid);
+
+  std::map<Dim, scipp::index> m_dims;
+  std::map<Dim, Variable> m_coords;
+  std::map<std::string, Variable> m_labels;
+  std::map<std::string, Variable> m_attrs;
+  std::map<std::string, detail::DatasetData, std::less<>> m_data;
+};
+
+/// Common functionality for other const-proxy classes.
+template <class Id, class Key> class ConstProxy {
+private:
+  struct make_item {
+    const ConstProxy *proxy;
+    auto operator()(const auto &item) const {
+      return std::pair<Key, ConstVariableSlice>(
+          item.first, detail::makeSlice(*item.second.first, proxy->slices()));
+    }
+  };
+
+public:
+  using key_type = Key;
+
+  ConstProxy(std::map<Key, std::pair<const Variable *, Variable *>> &&items,
+             const std::vector<std::pair<Slice, scipp::index>> &slices = {})
+      : m_items(std::move(items)), m_slices(slices) {
```

Comment:
Rather than initialising `m_items` "incorrectly" here and then purging in the loop(s) below. You could put the loop into a method, would be more like `m_items(sliceDimensions(items))`